tests/test_mypy.py

```diff
@@ -0,0 +1,212 @@
+
+    product_4 = Product(url=response.url)
+    product_4.canonicalUrl = response.url
+
+    product_5 = Product(url=response.url)
+    product_5.canonicalUrl = str(response.url)
+
+    product_6 = Product(url=response.url)
+    product_6.canonicalUrl = (
+        123  # E: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")
+    )
+
+
+@pytest.mark.mypy_testing
+def test_assignment_product_list():
+    product_list_1 = ProductList(url=response.url)
+    product_list_1.url = response.url
+
+    product_list_2 = ProductList(url=response.url)
+    product_list_2.url = str(response.url)
+
+    product_list_3 = ProductList(url=response.url)
+    product_list_3.url = 123  # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+    product_list_4 = ProductList(url=response.url)
+    product_list_4.canonicalUrl = response.url
+
+    product_list_5 = ProductList(url=response.url)
+    product_list_5.canonicalUrl = str(response.url)
+
+    product_list_6 = ProductList(url=response.url)
+    product_list_6.canonicalUrl = (
+        123  # E: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")
+    )
+
+
+@pytest.mark.mypy_testing
+def test_assignment_product_from_list():
+    product_from_list_1 = ProductFromList(url=response.url)
+    product_from_list_1.url = response.url
+
+    product_from_list_2 = ProductFromList(url=response.url)
+    product_from_list_2.url = str(response.url)
+
+    product_from_list_3 = ProductFromList(url=response.url)
+    product_from_list_3.url = (
+        123  # E: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")
+    )
+
+
+@pytest.mark.mypy_testing
+def test_assignment_product_variant():
+    product_variant_1 = ProductVariant(url=response.url)
+    product_variant_1.url = response.url
+
+    product_variant_2 = ProductVariant(url=response.url)
+    product_variant_2.url = str(response.url)
+
+    product_variant_3 = ProductVariant(url=response.url)
+    product_variant_3.url = (
+        123  # E: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")
+    )
+
+    product_variant_4 = ProductVariant(url=response.url)
+    product_variant_4.canonicalUrl = response.url
+
+    product_variant_5 = ProductVariant(url=response.url)
+    product_variant_5.canonicalUrl = str(response.url)
+
+    product_variant_6 = ProductVariant(url=response.url)
+    product_variant_6.canonicalUrl = (
+        123  # E: Incompatible types in assignment (expression has type "int", variable has type "Optional[str]")
+    )
+
+
+# NOTE: The tests below are marked to fail since the following error is
+# encountered in mypy:
+#
+#   "error: Skipping analyzing "web_poet": module is installed, but missing
+#   library stubs or py.typed marker"
```

Comment:
Here's a code snippet to demonstrate the **mypy** behavior when the stubs in `web-poet` would be available. The following code works because the URL classes are in the same script and not imported as a third-party package.

```python
from typing import Union
from typing_extensions import TypeGuard

import attrs


class Url:
    def __init__(self, url: Union[str, "Url"]):
        if not isinstance(url, (str, Url)):
            raise TypeError("err")
        self._url = str(url)

    def __str__(self) -> str:
        return self._url


class ResponseUrl(Url):
    pass


def is_url_obj(url: Union[str, Url]) -> TypeGuard[Url]:
    return True if isinstance(url, Url) else False


def url_to_str(value: Union[str, Url]) -> object:
    if is_url_obj(value):
        return str(value)
    return value


@attrs.define
class Image:
    url: str = attrs.field(converter=url_to_str)


url = ResponseUrl("123")
img = Image(url)

# error: Incompatible types in assignment (expression has type "int", variable has type "str")
img.url = 12

# error: Argument "url" to "Image" has incompatible type "int"; expected "Union[str, Url]"
f = Image(url=123)

# Second error does not pop out if this error is raised since it would render
# all classes coming from web_poet to be of type ``Any``.
# error: Skipping analyzing "web_poet": module is installed, but missing library stubs or py.typed marker

```